downloader/src/main/kotlin/VersionControlSystem.kt

```diff
@@ -58,10 +58,16 @@ abstract class VersionControlSystem {
          */
         fun forType(vcsType: String) = ALL.find { it.isApplicableType(vcsType) }
 
+        /**
+         * A map to cache the VersionControlSystem, if any, for previously queried URLs. This helps to speed up queries
+         * for the same URLs as running e.g. "hg identify" takes several seconds on each.
+         */
+        private val urlToVcsMap = mutableMapOf<String, VersionControlSystem?>()
+
         /**
          * Return the applicable VCS for the given [vcsUrl], or null if none is applicable.
          */
-        fun forUrl(vcsUrl: String) = ALL.find { it.isApplicableUrl(vcsUrl) }
+        fun forUrl(vcsUrl: String) = urlToVcsMap.getOrPut(vcsUrl) { ALL.find { it.isApplicableUrl(vcsUrl) } }
```

Comment:
This does not work if `ALL.find {...}` returns null, because `getOrPut` does despite of the docs not actually check if the map contains an entry but only if the value is `!= null`. Therefore `defaultValue` is always executed for those `vcsUrl`s.

If I change the implementation to e.g. this it works as expected and significantly reduces the runtime for the tests I'm currently working on:

```kotlin
        fun forUrl(vcsUrl: String) = if (urlToVcsMap.containsKey(vcsUrl)) {
            urlToVcsMap[vcsUrl]
        } else {
            ALL.find { it.isApplicableUrl(vcsUrl) }.also {
                urlToVcsMap[vcsUrl] = it
            }
        }
```

Also see: [KT-21392](https://youtrack.jetbrains.com/issue/KT-21392) and [PR #1710](https://github.com/JetBrains/kotlin/pull/1710)

Reply:
Thanks for pointing that out!